### PR TITLE
[Calyx] Use explicit in- and output port lists in ComponentOp::build

### PR DESCRIPTION
**Pull request by mortbopet**
If the ArrayRef ports argument contains an unordered list of ports, the first numInPorts ports are registered as the input ports, and the remainder as outputs.
I.e. the following list of ports:

```c++
SmallVector<calyx::ComponentPortInfo> ports;
ports.push_back({rewriter.getStringAttr("clk"), rewriter.getI1Type(), calyx::PortDirection::INPUT});
ports.push_back({rewriter.getStringAttr("reset"), rewriter.getI1Type(), calyx::PortDirection::INPUT});
ports.push_back({rewriter.getStringAttr("done"), rewriter.getI1Type(), calyx::PortDirection::OUTPUT}); // <--
ports.push_back({rewriter.getStringAttr("go"), rewriter.getI1Type(), calyx::PortDirection::INPUT});
```

will currently (incorrectly) create a component with the following signature:

```mlir
calyx.component @main(%clk: i1, %reset: i1, %done: i1) -> (%go: i1) {
   ...
}
```

The commit introduces explicit input- and output port lists.
The `Direction` of the ports within the `calyx::ComponentPortInfo` structs passed to ComponentOp::build are ignored; this aligns with the remainder of the logic surrounding port direction inference in `CalyxOps.cpp`, wherein the first `nInputPorts` are regarded as inputs, and the remainder as outputs _regardlessly_ of what `Direction` was set for the port when building the component.

### CODE REVIEW

**Comment by mortbopet** [comment]
Pinging @cgyurgyik

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:311]: I see two alternatives to this.

1. Separately track port directions as an attribute, similar to [FIRRTL](https://github.com/llvm/circt/blob/ee457605169c16b1ce1400201e66d06d723f282b/include/circt/Dialect/FIRRTL/FIRRTLOps.h#L40-L53). We would no longer need this pre-condition that inputs are listed before outputs.

2.  Keeping this as a single list of `ports`, and then partitioning the ports by direction `{inputs, outputs}`. 

Something like:
```
 auto portIt = llvm::partition(ports, [](auto port){ return port.direction == INPUT; });
 uint64_t numInPorts = std::distance(ports.begin(), portIt);
 ...
```

Reasoning: The current changes redundantly lists input and output ports, which should already be identified by the `ComponentPortInfo::PortDirection`.

WDYT?

**Review by cgyurgyik — COMMENTED** [review]
Thank you for catching this! I've left two possible alternatives. Option 1 requires slightly more work but results in a safer solution. Option 2 is a slight variation of your suggestion.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp:311]: Regarding option 2, if the partition is deterministic in terms of not reordering the inputs and the outputs themselves, then that would work. But option 1 would get closer to solving the actual issue, being that port directionality throughout the ComponentOp functions is being tracked as information based on the provided arguments during construction and the deduced # of inputs.

Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp]: 5b646b9 implements option 1.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:380]: Nit: Save this as a boolean, e.g. `bool isInput = port.direction == Direction::Input`.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:356]: Please document this. 

Another option is using [SmallVectorImpl::append](https://llvm.org/doxygen/classllvm_1_1SmallVectorImpl.html#acd79f15ee5c01bfc3e2393233f40b764), which will also reserve space.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:374]: Good catch. I meant `llvm::stable_partition` in my previous comment, but your current solution is better.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:49]: `const` not necessary.
Tiny nit: please use more descriptive name such as `numDirections`.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:52]: did `directions.size()` change here, or can we use the variable above?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:62]: No `const` necessary.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:184]: Nit (feel free to ignore): curly braces not necessary if one-liner. However, there is quite a lot going on here :-).

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:280]: Nit (feel free to ignore): I don't have a computer available to try myself, but it *may* be possible to do something like `ports.append({inPorts, outPorts})`?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:391]: Can you explain why this is necessary? I'm not familiar.

**Review by cgyurgyik — COMMENTED** [review]
This is awesome!! Thank you! I've left a few comments, mostly nit picks.

File: include/circt/Dialect/Calyx/CalyxOps.h
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxOps.h:60]: Grammar nit: `a vector of directions`.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:66]: Good use of `reserve`. Perhaps we can save bit width to duplicate calls?

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp:280]: Unfortunately does not seem to be the case. I was hoping for functionality like that (alongside the `concat` function i added) would be present in STLExtras.h

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp:391]: Thank you for catching that! That's a leftover from some debugging i did.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp:66]: Could you elaborate? Not sure what you mean here wrt. duplicate calls.

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp:66]:
```
auto bitWidth = value.getBitWidth();
result.reserve(bitWidth);
...
for (size_t i = 0, e = bitWidth; i != e; ++i)
  ...
```

Comment (mortbopet) [reply lib/Dialect/Calyx/CalyxOps.cpp:66]: Ahh, sorry - totally misread your message. I interpreted "save bit width" as in somehow reducing the # of bits stored in the IntegerAttr vector.

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp]: No problem, I should've been more explicit. Overall LGTM, I'm going to add @mikeurbach for a final review. Thanks again for the contribution @mortbopet ! 

In the mean time, if you're interested in committing this yourself after Mike's approval, please take a look at the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access).

File: include/circt/Dialect/Calyx/CalyxOps.h
Comment (mikeurbach) [inline include/circt/Dialect/Calyx/CalyxOps.h:50]: Style comment: I somewhat prefer a static method like `Direction::get` rather than a function in the namespace like `direction::get`. The former is more in line with how MLIR provides helpers for things like attributes (e.g. `ArrayAttr::get`). Up to you though.

File: include/circt/Dialect/Calyx/CalyxOps.h
Comment (mikeurbach) [inline include/circt/Dialect/Calyx/CalyxOps.h:54]: Uber nit: but if you're going to name the parameters in the header here, I'd recommend using the more descriptive names that you gave them in the cpp file rather than just `a` and `b`.

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mikeurbach) [inline lib/Dialect/Calyx/CalyxOps.cpp:359]: If this is just an internal helper, I think you can (and should?) mark it `static`.

**Review by mikeurbach — APPROVED** [review]
Couple minor nitpicks, but this looks good to me. If you're happy with this and are awaiting commit permissions, let us know and someone can merge it for you.

File: include/circt/Dialect/Calyx/CalyxOps.h
Comment (mortbopet) [reply include/circt/Dialect/Calyx/CalyxOps.h:50]: I understand - I didn't pay careful attention to this code since it [essentially is a copy of what is currently done in FIRRTL](https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/FIRRTLOps.h#L38).

My hunch would be to leave it as is. This would put it in the scope of anyone in the future refactoring this and the FIRRTL code to a single, shared implementation.

File: include/circt/Dialect/Calyx/CalyxOps.h
Comment (mortbopet) [reply include/circt/Dialect/Calyx/CalyxOps.h:54]: As above.

**Comment by mortbopet** [comment]
@mikeurbach Thank you for the review. I'll be applying for commit permissions as soon as possible, however in the meantime, I'd be happy if someone could merge it for me.

**Comment by mikeurbach** [comment]
:+1: thanks for the contribution. I just triggered the CI checks (had to be done manually, but will be automatic once you've had a PR merged). If they pass I'll go ahead and click merge.

**Pull request merged**
